### PR TITLE
Move basic cursor

### PR DESCRIPTION
**Pull request by alschmiedt**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Proposed Changes
1.  Move BasicCursor to core
2. Refactor BasicCursor to make it easier to modify
3. Add a LineCursor to demos

### Reason for Changes
1. Basic cursor seems to be more useful than I originally thought so I think it should be in core. This will allow other cursors to easily extend it such as the TabNavigateCursor.
3. I think this is a useful demo cursor and shows how to modify the BasicCursor easily. 

### Test Coverage

Tested on:

### CODE REVIEW

File: demos/keyboard_nav/line_cursor.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+ */
+'use strict';
+
+
+/**
+ * Class for a line cursor.
+ * This will allow the user to get to all nodes in the AST by hitting next or
+ * previous.
+ * @constructor
+ * @extends {Blockly.LineCursor}
```
Comment (samelhusseini) [inline demos/keyboard_nav/line_cursor.js]: Extends itself? I think you mean extends BasicCursor

File: demos/keyboard_nav/line_cursor.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      location.type == Blockly.NEXT_STATEMENT) {
+    isValid = true;
+  } else if (type == Blockly.ASTNode.types.NEXT) {
+    isValid = true;
+  }
+  return isValid;
+};
+
+/**
+ *
```
Comment (samelhusseini) [inline demos/keyboard_nav/line_cursor.js]: Add jsdoc.

File: demos/keyboard_nav/line_cursor.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+};
+
+/**
+ * Decides what nodes to traverse and which ones to skip. Currently, it
+ * skips output, stack and workspace nodes.
+ * @param {Blockly.ASTNode} node The AST node to check whether it is valid.
+ * @return {boolean} True if the node should be visited, false otherwise.
+ * @private
+ */
+Blockly.LineCursor.prototype.validNode_ = function(node) {
```
Comment (samelhusseini) [inline demos/keyboard_nav/line_cursor.js]: Can we come up with a better name for validNode_ and validInNode_, its hard to tell what the difference is without reading comments.

File: demos/keyboard_nav/line_cursor.js (hunk 1 above)
Comment (alschmiedt) [reply demos/keyboard_nav/line_cursor.js]: Woops, fixed!

File: demos/keyboard_nav/line_cursor.js (hunk 3 above)
Comment (alschmiedt) [reply demos/keyboard_nav/line_cursor.js]: We can but I am struggling haha.  validLineNode and validInLineNode?

File: demos/keyboard_nav/line_cursor.js (hunk 2 above)
Comment (alschmiedt) [reply demos/keyboard_nav/line_cursor.js]: done!

File: demos/keyboard_nav/line_cursor.js (hunk 3 above)
Comment (samelhusseini) [reply demos/keyboard_nav/line_cursor.js]: Sounds good.

**Pull request merged**